File: Application/src/main/java/com/shareyourproxy/api/domain/factory/UserTypeAdapterFactory.java
package com.shareyourproxy.api.domain.factory;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.shareyourproxy.api.domain.model.User;

import static com.shareyourproxy.util.ObjectUtils.buildFullName;

/**
 * Created by Evan on 11/3/15.
 */
public class UserTypeAdapterFactory extends CustomTypeAdapterFactory<User> {
    public UserTypeAdapterFactory() {
        super(User.class);
    }

    @Override
    protected void beforeWrite(User source, JsonElement toSerialize) {
        JsonObject custom = toSerialize.getAsJsonObject();
        custom.remove("fullName");
    }

    @Override
    protected void afterRead(JsonElement deserialized) {
        if (deserialized.isJsonObject()) {
            JsonObject obj = deserialized.getAsJsonObject();
            JsonElement firstName = obj.get("first");
            JsonElement lastName = obj.get("last");
            String first = firstName == null ? "" : firstName.getAsString();
            String last = lastName == null ? "" : lastName.getAsString();
            obj.add("fullName", new JsonPrimitive(buildFullName(first, last)));
        }
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/domain/factory/UserTypeAdapterFactory.kt
package com.shareyourproxy.api.domain.factory

import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.util.StringUtils.buildFullName

/**
 * The web service actually doesn't return a fullname param, we make/strip it here from the user object.
 */
object UserTypeAdapterFactory : CustomTypeAdapterFactory<User>(User::class.java) {

    override fun beforeWrite(source: User, toSerialize: JsonElement) {
        val custom = toSerialize.asJsonObject
        custom.remove("fullName")
    }

    override fun afterRead(deserialized: JsonElement) {
        if (deserialized.isJsonObject) {
            val obj = deserialized.asJsonObject
            val firstName = obj.get("first")
            val lastName = obj.get("last")
            val first = if (firstName == null) "" else firstName.asString
            val last = if (lastName == null) "" else lastName.asString
            obj.add("fullName", JsonPrimitive(buildFullName(first, last)))
        }
    }
}
